placement/service/placementservice.go

```diff
@@ -233,39 +222,112 @@ func (ps placementService) findReplaceHost(s placement.Snapshot, candidateHosts
-	ph := algo.NewPlacementHelper(ps.options, p)
+func groupHostsByConflict(hostsSortedByConflicts []sortableThing, allowConflict bool) [][]placement.Host {
+	sort.Sort(sortableThings(hostsSortedByConflicts))
+	var groups [][]placement.Host
+	lastSeenConflict := -1
+	for _, host := range hostsSortedByConflicts {
+		if !allowConflict && host.value > 0 {
+			break
+		}
+		if host.value > lastSeenConflict {
+			lastSeenConflict = host.value
+			groups = append(groups, []placement.Host{})
+		}
+		if lastSeenConflict == host.value {
+			groups[len(groups)-1] = append(groups[len(groups)-1], host.thing.(placement.Host))
+		}
+	}
+	return groups
+}
 
-	rackLens := make(rackLens, 0, len(rackHostMap))
-	for rack := range rackHostMap {
-		rackConflicts := 0
-		for _, shard := range leaving.Shards() {
-			if !ph.HasNoRackConflict(shard, leaving, rack) {
-				rackConflicts++
-			}
+func fillWeight(groups [][]placement.Host, targetWeight int) ([]placement.Host, int) {
+	var (
+		result       []placement.Host
+		hostsInGroup []placement.Host
+	)
+	for _, group := range groups {
+		hostsInGroup, targetWeight = knapsack(group, targetWeight)
+		result = append(result, hostsInGroup...)
+		if targetWeight <= 0 {
+			break
 		}
 
-		rackLens = append(rackLens, rackLen{rack: rack, len: rackConflicts})
 	}
-	sort.Sort(rackLens)
-	if !ps.options.LooseRackCheck() {
-		rackLens = filterConflictRacks(rackLens)
+	return result, targetWeight
+}
+
+func knapsack(hosts []placement.Host, targetWeight int) ([]placement.Host, int) {
+	totalWeight := 0
+	for _, host := range hosts {
+		totalWeight += int(host.Weight())
+	}
+	if totalWeight <= targetWeight {
+		return hosts[:], targetWeight - totalWeight
+	}
+	// totalWeight > targetWeight, there is a combination of hosts to meet targetWeight for sure
+	// we do dp until totalWeight rather than targetWeight here because we need to cover the targetWeight
+	// which is a little bit different than the knapsack problem that goes
+	weights := make([]int, totalWeight+1)
+	combination := make([][]placement.Host, totalWeight+1)
+
+	for _, host := range hosts {
+		for i := totalWeight; i >= 1; i-- {
+			weight := int(host.Weight())
+			if i-weight < 0 {
+				continue
+			}
+			newWeight := weights[i-weight] + weight
+			if newWeight > weights[i] {
+				weights[i] = weights[i-weight] + weight
+				combination[i] = append(combination[i-weight], host)
+			} else if newWeight == weights[i] {
+				// if can reach same weight, find a combination with least number of hosts
+				if len(combination[i-weight])+1 < len(combination[i]) {
+					combination[i] = append(combination[i-weight], host)
+				}
+			}
+		}
 	}
-	if len(rackLens) > 0 {
-		return rackHostMap[rackLens[0].rack][0], nil
+	for i := targetWeight; i <= totalWeight; i++ {
+		if weights[i] >= targetWeight {
+			return combination[i], targetWeight - weights[i]
+		}
 	}
-	return nil, errNoValidHost
+
+	// will never reach here
+	return nil, 0
```

Comment:
If we should never reach here, should we log an error or even panic to catch the bug etc?